sw/device/lib/dif/dif_template.h.tpl

```diff
@@ -0,0 +1,248 @@
+ *
+ * This enum may be used instead of a `bool` when describing an enabled/disabled
+ * state.
+ */
+typedef enum dif_<ip>_toggle {
+  /**
+   * The "enabled" state.
+   */
+  kDif<IP>ToggleEnabled,
+  /**
+   * The "disabled" state.
+   */
+  kDif<IP>ToggleDisabled,
+} dif_<ip>_toggle_t;
+
+/**
+ * Hardware instantiation parameters for <peripheral>.
+ *
+ * This struct describes information about the underlying hardware that is
+ * not determined until the hardware design is used as part of the a toplevel.
+ */
+typedef struct dif_<ip>_param {
+  /**
+   * The base address for the <peripheral> hardware registers.
+   */
+  mmio_region_t base_addr;
+
+  // Other fields, if necessary.
+} dif_<ip>_param_t;
+
+/**
+ * Runtime configuration for <peripheral>.
+ *
+ * This struct describes runtime information for one-time configuration of the
+ * hardware.
+ */
+typedef struct dif_<ip>_config {
+  // Fields, if necessary.
+} dif_<ip>_config_t;
+
+/**
+ * A handle to <peripheral>.
+ *
+ * This type should be treated as opaque by users.
+ */
+typedef struct dif_<ip> {
+  dif_param_<ip>_t params;
+
+  // Other fields, if necessary.
+} dif_<ip>_t;
+
+/**
+ * The result of a <peripheral> operation.
+ */
+typedef enum dif_<ip>_result {
+  /**
+   * Indicates that the operation succeeded.
+   */
+  kDif<IP>Ok = 0,
+  /**
+   * Indicates some unspecified failure.
+   */
+  kDif<IP>Error = 1,
+  /**
+   * Indicates that some parameter passed into a function failed a
+   * precondition.
+   *
+   * When this value is returned, no hardware operations occured.
+   */
+  kDif<IP>BadArg = 2,
+} dif_<ip>_result_t;
+
+/**
+ * A <peripheral> interrupt request type.
+ */
+typedef enum dif_<ip>_irq {
+  // Your IRQs here!
+} dif_<ip>_irq_t;
+
+/**
+ * A snapshot of the enablement state of the interrupts for <peripheral>.
+ *
+ * This is an opaque type, to be used with the `dif_<ip>_irq_disable_all()` and
+ * `dif_<ip>_irq_restore_all()` functions.
+ */
+typedef uint32_t dif_<ip>_irq_snapshot_t;
+
+/**
+ * Creates a new handle for <peripheral>.
+ *
+ * This function does not actuate the hardware.
+ *
+ * @param params Hardware instantiation parameters.
+ * @param handle Out param for the initialized handle.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_<ip>_result_t dif_<ip>_init(dif_<ip>_params_t params, dif_<ip>_t *handle);
+
+/**
+ * Configures <peripheral> with runtime information.
+ *
+ * This function should need to be called once for the lifetime of `handle`.
+ *
+ * @param handle A <peripheral> handle.
+ * @param config Runtime configuration parameters.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_<ip>_result_t dif_<ip>_configure(const dif_<ip>_t* handle,
+                                     dif_<ip>_config_t config);
+
+/**
+ * Returns whether a particular interrupt is currently pending.
+ *
+ * @param handle A <peripheral> handle.
+ * @param irq_type An interrupt type.
```

Comment:
I don't think `irq_type` and `interrupt type` is the right terminology to use (reminds me of hardware and software interrupts). IMHO `irq` should be enough. (Other usages will need to be updated so let's make sure before making a change).

Reply:
Unresolved for now to give people a chance to object.

Reply:
No one has objected, so I'm resolving going with Alphan's suggestion.